backend/controllers/authController.js

```diff
@@ -0,0 +1,220 @@
+
+        if (!isMatch) {
+            console.log(`Login attempt failed: Incorrect password for user with email ${email}.`);
+            return res.status(401).json({ msg: 'Invalid credentials' });
+        }
+
+        // Create a JWT token payload
+        const payload = { user: { id: user.id } };
+        // Generate JWT token
+        const token = jwt.sign(payload, process.env.JWT_SECRET, { expiresIn: '12h' });
+        // Return token and success message
+        return res.status(200).json({
+            token,
+            msg: 'Logged in successfully',
+            user: {
+                id: user.id,
+                first_name: user.first_name,
+                last_name: user.last_name,
+                email: user.email,
+                isDonor: user.isDonor,
+                isStudent: user.isStudent
+            }
+        });
+    } catch (err) {
+        console.error(`Error during login for email: ${email}: ${err.message}`);
+        return res.status(500).json({ msg: 'Server error' });
+    }
+};
+
+const forgotPassword = async(req, res) => {
+    const { email } = req.body;
+    try {
+        const user = await UserModel.findOne({ email });
+
+        if (!user) {
+            return res.status(404).json({ msg: 'User does not exist' });
+        }
+
+        const resetCode = Math.floor(100000 + Math.random() * 900000); // 6-digit reset code
+
+        // send reset password email
+        const userDetail = {
+            name: user.first_name + ' ' + user.last_name,
+            email: user.email
+        }
+        try {
+            const emailResponse = await sendMail(userDetail,
+                'Password Reset',
+                `Your password reset code is: ${resetCode}`);
+            user.passwordResetCode = resetCode;
+            await user.save();
+
+            return res.status(emailResponse.status).json({ message: emailResponse.message });
+        } catch (error) {
+            return res.status(error.status || 500).json({ msg: "Error sending email", error: error.message });
+        }
+    } catch (err) {
+        console.error(`Error during forgot password for email: ${email}: ${err.message}`);
+        return res.status(500).json({ msg: err.message });
+    }
+};
+
+const verifyResetCode = async(req, res) => {
+    const errors = validationResult(req);
+    if (!errors.isEmpty()) {
+        return res.status(400).json({ errors: errors.array() });
+    }
+    const { email, verificationCode } = req.body;
+    try {
+        // Find user by email and verification code
+        const user = await UserModel.findOne({ email, passwordResetCode: verificationCode });
+        if (!user) {
+            console.log(`Verification attempt failed: Invalid verification code for email ${email}.`);
+            return res.status(401).json({ msg: 'Invalid verification code' });
+        }
+        // If the code is correct, return success
+        return res.status(200).json({ msg: 'Code verified successfully' });
```

Comment:
When user forget the password I ask for email. In that logic, I am checking if user email exists, if not I don't allow that user to proceed until they enter valid email. So here I am just verifying code user enter on UI is valid or not. So that's why 401.

Reply:
Oh, okay.